### ISSUE TITLE
Use localStorage for Google Analytics tracking when available

### ISSUE THREAD

**Issue by davidmurdoch** [h5bp/html5-boilerplate#1444]
TL;DR:

``` js
(function(b,o,i,l,e,r){b.GoogleAnalyticsObject=l;b[l]||(b[l]=
function(){(b[l].q=b[l].q||[]).push(arguments)});b[l].l=+new Date;
e=o.createElement(i);r=o.getElementsByTagName(i)[0];
e.src='//www.google-analytics.com/analytics.js';
r.parentNode.insertBefore(e,r)}(window,document,'script','ga'));
ga('create','UA-XXXXX-X',{'storage': 'none','clientId':localStorage.getItem('gaClientId')});
ga(function(t){localStorage.setItem('gaClientId',t.get('clientId'));});
ga('send','pageview');
```

---

**The Source:**
http://stackoverflow.com/questions/4502128/convert-google-analytics-cookies-to-local-session-storage/19207035?noredirect=1#19207035

**Google Analytics Docs:**
https://developers.google.com/analytics/devguides/collection/analyticsjs/domains#disableCookies

We could use `Modernizer.localstorage` to check for `localStorage` support and fallback to cookies if its not available. Though I'm not sure if we want to lock in Modernizr as a dependency.

**Why?**
Because Google doesn't need to send their cookie to your server for every single request to your domain (or theirs, for that matter).

**Comment by caesarsol** [comment]
Hello, have you read this SO comment?

http://stackoverflow.com/questions/4502128/convert-google-analytics-cookies-to-local-session-storage/19207035#comment-44767913

I'd be curious to know what you all think.

**Comment by philipwalton** [comment]
@caesarsol I think that's a really bad idea. As I described in [my comment](https://github.com/h5bp/html5-boilerplate/issues/1444#issuecomment-57235840), cookies and localStorage don't have the same restrictions, so swapping them out for every single script that runs on the page is extremely risky.

**Comment by caesarsol** [comment]
hello @philipwalton, thanks for the response but maybe I explained poorly, i was referring to this comment by SO user `smhmic`:

>  This might violate GA TOS! Here is a secondhand quote from a GA Team member, taken from this article: "Using HTTP State Management mechanisms" (read: localStorage) "to propagate cookie state is a circumvention of our privacy safeguards. Doing so violates the Google Analytics Terms of Service". My interpretation of this is that GA employs cookies and not localStorage because more users are familiar with the concept of cookies and how to clear them; thus, GA's use of cookies is a privacy feature. –  smhmic

**Comment by philipwalton** [comment]
> Using HTTP State Management mechanisms" (read: localStorage) to propagate cookie state is a circumvention of our privacy safeguards. Doing so violates the Google Analytics Terms of Service

Hmmm, I don't think this is true. There are opt-out features that GA provides (e.g. Chrome extensions) that don't rely on the implementor using cookies. I think the point of this section of the TOS is that you can't create a mechanism by which someone who's using an official "do not track" extension will *still* be tracked.

I can look into it further and I'll update this thread is my assumptions turn out to be false.

**Comment by davidmurdoch** [comment]
**Update:**

It is not against TOS to use localStorage to store the ClientID; it is now officially supported by Google: https://developers.google.com/analytics/devguides/collection/analyticsjs/cookies-user-id#using_localstorage_to_store_the_client_id

Note: if you have to support (extremely) old browsers (like iOS5 and FF4) their example snippet may fail (see: https://github.com/Modernizr/Modernizr/blob/master/feature-detects/storage/localstorage.js).
